File: src/app/mcp/kimi.ts
import { mapMcpServers, type McpServerSpec } from "@keepdeck/plugin-api";

/**
 * kimi's dialect — the one that lives host-side rather than in its plugin.
 *
 * Every other CLI takes its servers through argv or env, which is a hook's
 * business. kimi has neither: its loader reads `<cwd>/.kimi-code/mcp.json`, so
 * the delivery is a FILE, and files in a pane's working directory are written
 * by the host (plugins have no such capability, and the ordering against
 * worktree teardown is the worktree owner's invariant). The same split codex
 * already has for skills: the host plants, the plugin contributes nothing.
 *
 * Host-side is the SETTLED place for the format, not a priced-shut door: a
 * plugin-declared one was designed and rejected — it would have carried the
 * shape as wire strings, trading a compile error for a silent mismatch the day
 * a second dialect arrives. What the host owes instead is that no flow names
 * an agent: it asks [`mcpFileRenderer`], and the Rust side asks its own view
 * (`mcp/kimi.rs`). A second dialect adds a module beside this one.
 */
const KIMI_AGENT = "kimi";

/** Renders one pane's servers into the body its CLI reads from disk. */
export type McpFileRenderer = (servers: readonly McpServerSpec[]) => string;

/**
 * How this agent's servers reach it: a renderer when the CLI is fed by a FILE
 * in the pane's cwd, `null` when it takes them on argv.
 *
 * The question the injection flow asks, so it never names an agent. Everything
 * that follows from the answer — that a shared cwd holds one file and so one
 * secret, that nothing rides argv — follows from file delivery as a CLASS, not
 * from which CLI happens to be file-fed today.
 */
export function mcpFileRenderer(agentType: string): McpFileRenderer | null {
  return agentType === KIMI_AGENT ? kimiMcpConfig : null;
}

/** The `mcp.json` body for one pane, in kimi's own shape (`mcpServers`, keyed
 * by name — the format claude's config also uses, which is why kimi calls the
 * repo-root variant "Claude-compatible").
 *
 * A remote server names its `transport` outright — kimi's loader discriminates
 * on that field — and carries the bearer token as `bearerTokenEnvVar`, kimi's
 * own token-from-env field, so the value stays in the pane's environment. A
 * stdio server's `envPassthrough` renders nothing: kimi hands its MCP children
 * its whole environment (probe-verified on 0.31.1). */
export function kimiMcpConfig(servers: readonly McpServerSpec[]): string {
  const mcpServers = Object.fromEntries(
    mapMcpServers<[string, Record<string, unknown>]>(servers, {
      stdio: (server) => [server.name, { command: server.command, args: server.args }],
      http: (server) => [
        server.name,
        {
          transport: "http",
          url: server.url,
          ...(server.headers ? { headers: server.headers } : {}),
          ...(server.bearerTokenEnv
            ? { bearerTokenEnvVar: server.bearerTokenEnv }
            : {}),
        },
      ],
    }),
  );
  return `${JSON.stringify({ mcpServers }, null, 2)}\n`;
}

File: src-tauri/src/mcp/kimi.rs
//! kimi's view of a pane cwd — where its MCP config lives, and nothing else.
//!
//! The same division [`crate::skills::opencode`] makes for staging: the arming
//! flow says WHEN a config is planted or taken back and what makes it provably
//! ours; this says WHAT it is called. A second file-fed CLI adds a view beside
//! this one instead of editing the flow every dialect shares.
//!
//! Only the two names below are kimi's. The marker that claims the directory
//! is KeepDeck's convention, identical for every dialect, so it stays with the
//! flow.

use std::path::{Path, PathBuf};

/// The directory kimi reads its MCP config from, inside the pane's cwd.
const DIR: &str = ".kimi-code";
/// The file it reads there.
const CONFIG: &str = "mcp.json";

/// One pane cwd, seen the way kimi sees it.
pub(super) struct View {
    dir: PathBuf,
    config: PathBuf,
}

impl View {
    pub(super) fn at(cwd: &Path) -> Self {
        let dir = cwd.join(DIR);
        Self {
            config: dir.join(CONFIG),
            dir,
        }
    }

    /// The planted directory. Git is kept blind to it by this name, and a
    /// non-directory here is the user's own — the flow decides what to do
    /// about that, the name is ours.
    pub(super) fn dir(&self) -> &Path {
        &self.dir
    }

    /// The config itself.
    pub(super) fn config(&self) -> &Path {
        &self.config
    }

    /// The directory's name alone, for the `info/exclude` line and for the
    /// refusals that have to name what is actually there.
    pub(super) fn dir_name() -> &'static str {
        DIR
    }

    /// The config's name alone, for the same reason.
    pub(super) fn config_name() -> &'static str {
        CONFIG
    }
}
